hatasmota/entity.py

```diff
@@ -98,27 +104,65 @@ def __init__(self, **kwds: Any):
         self._on_availability_callback: Callable[
             [bool], Coroutine[Any, Any, None]
         ] | None = None
+        self.uses_deep_sleep = False
+        self.deep_sleep_interval = None
+        self.last_up=0
+        self.available=None
         super().__init__(**kwds)
 
     def get_availability_topics(self) -> dict:
         """Return MQTT topics to subscribe to for availability state."""
 
         async def availability_message_received(msg: ReceiveMessage) -> None:
             """Handle a new received MQTT availability message."""
+            last_up_retain=self.last_up
             if msg.payload == self._cfg.availability_online:
+                if self.last_up and self.deep_sleep_interval is None:
+                    self.deep_sleep_interval = int(time.time() - self.last_up)
                 await self.poll_status()
+            else:
+                self.last_up=time.time()
+            available=self.available
+            if msg.payload == self._cfg.availability_online:
+                available=True
+            if msg.payload == self._cfg.availability_offline and not(self.uses_deep_sleep):
+                if not(self.uses_deep_sleep):
+                    available=False
+                else:
+                    _LOGGER.debug("inhibit deep sleep %s", msg.topic)
             if not self._on_availability_callback:
+                self.available=available
                 return
-            if msg.payload == self._cfg.availability_online:
-                await self._on_availability_callback(True)
-            if msg.payload == self._cfg.availability_offline:
-                await self._on_availability_callback(False)
-
+            if self.available != available: 
+                await self._on_availability_callback(available)
+                self.available=available
+
+        async def sleep_state_message_received(msg: ReceiveMessage) -> None:
+            """Handle state messages to indicate deep sleep."""
+            #try:
+            #    payload = json.loads(msg.payload)
+            #except json.decoder.JSONDecodeError:
+            #    return
+            _LOGGER.debug("sleep state %s -> %s", msg.topic, msg.payload)
+            state = get_value_by_path(msg.payload, ["StatusPRM","RestartReason"])
+            if state is not None:
+                state=str(state).lower()
+                if state.startswith('deep sleep'):
+                    if not(self.uses_deep_sleep):
+                        _LOGGER.debug("switching to deep sleep mode %s", msg.topic)
+                        self.deep_sleep_interval=None
+                        self.uses_deep_sleep=True
```

Comment:
Instead of detecting a device's support for deep sleep when it wakes up from deep sleep, we should detect it from the device's discovery message: <https://github.com/arendst/Tasmota/pull/19134/files#diff-fe970b21f48758677cb799d64f0ac866cbbe0aa48eb3e0f6aaa52c22ef73c735R189>
Also